README.md
# Network_type_reachability

This plugin allows Flutter apps to detect network changes. You can know the detailed mobile network types, such as 2G, 3G, 4G, 5G. This plugin is suitable for iOS and Android, it can also detect if the internet connection status is real or not working and even if it is unstable, it allows readings in a single call by creating a listener

This library is based on and inspired by the library "[flutter_reachability](https://pub.dev/packages/flutter_reachability)"

This library uses the following dependencies but it is not necessary to place it in pubspec.yalm because everything is already included in the "Network_type_reachability" library

- [permission_handler: 9.2.0](https://pub.dev/packages/permission_handler/versions/9.2.0)<br>
- [dart_ping: ^6.1.2](https://pub.dev/packages/dart_ping/versions/6.1.2)<br>
- [dart_ping_ios: ^1.1.0](https://pub.dev/packages/dart_ping_ios/versions/1.1.0)

Next, the necessary configurations for the correct operation of "Network_type_reachability" will be shown, but it is recommended to take a look at the original packages.

***

To use the version prior to NullSafety use the version [1.2.1](https://pub.dev/packages/network_type_reachability/versions/1.2.1)

***
## Permission_handler settings

<details close>
<summary>Android</summary>


**Upgrade pre 1.12 Android projects**

Since version 4.4.0 this plugin is implemented using the Flutter 1.12 Android plugin APIs. Unfortunately this means App developers also need to migrate their Apps to support the new Android infrastructure. You can do so by following the [Upgrading pre 1.12 Android projects](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects) migration guide. Failing to do so might result in unexpected behaviour. Most common known error is the permission\_handler not returning after calling the `.request()` method on a permission.

**AndroidX**

As of version 3.1.0 the permission\_handler plugin switched to the AndroidX version of the Android Support Libraries. This means you need to make sure your Android project is also upgraded to support AndroidX. Detailed instructions can be found [here](https://flutter.dev/docs/development/packages-and-plugins/androidx-compatibility).

The TL;DR version is:

1.  Add the following to your "gradle.properties" file:

    android.useAndroidX=true
    android.enableJetifier=true
    

copied to clipboard

1.  Make sure you set the `compileSdkVersion` in your "android/app/build.gradle" file to 31:

    android {
      compileSdkVersion 31
      ...
    }
    

copied to clipboard

1.  Make sure you replace all the `android.` dependencies to their AndroidX counterparts (a full list can be found [here](https://developer.android.com/jetpack/androidx/migrate)).

Add permissions to your `AndroidManifest.xml` file. There's a `debug`, `main` and `profile` version which are chosen depending on how you start your app. In general, it's sufficient to add permission only to the `main` version. [Here](https://github.com/Baseflow/flutter-permission-handler/blob/master/permission_handler/example/android/app/src/main/AndroidManifest.xml)'s an example `AndroidManifest.xml` with a complete list of all possible permissions.
</details>

<details close>
<summary>Ios</summary>

Add permission to your `Info.plist` file. [Here](https://github.com/Baseflow/flutter-permission-handler/blob/master/permission_handler/example/ios/Runner/Info.plist)'s an example `Info.plist` with a complete list of all possible permissions.

> IMPORTANT: You will have to include all permission options when you want to submit your App. This is because the `permission_handler` plugin touches all different SDKs and because the static code analyser (run by Apple upon App submission) detects this and will assert if it cannot find a matching permission option in the `Info.plist`. More information about this can be found [here](https://github.com/Baseflow/flutter-permission-handler/issues/26).

The permission\_handler plugin use [macros](https://github.com/Baseflow/flutter-permission-handler/blob/master/permission_handler/ios/Classes/PermissionHandlerEnums.h) to control whether a permission is enabled.

You must list permission you want to use in your application :

1.  Add the following to your `Podfile` file:
    
        post_install do |installer|
          installer.pods_project.targets.each do |target|
            target.build_configurations.each do |config|
              ... # Here are some configurations automatically generated by flutter
        
              # You can enable the permissions needed here. For example to enable camera
              # permission, just remove the `#` character in front so it looks like this:
              #
              # ## dart: PermissionGroup.camera
              # 'PERMISSION_CAMERA=1'
              #
              #  Preprocessor definitions can be found in: https://github.com/Baseflow/flutter-permission-handler/blob/master/permission_handler/ios/Classes/PermissionHandlerEnums.h
              config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
                '$(inherited)',
        
                ## dart: PermissionGroup.calendar
                # 'PERMISSION_EVENTS=1',
        
                ## dart: PermissionGroup.reminders
                # 'PERMISSION_REMINDERS=1',
        
                ## dart: PermissionGroup.contacts
                # 'PERMISSION_CONTACTS=1',
        
                ## dart: PermissionGroup.camera
                # 'PERMISSION_CAMERA=1',
        
                ## dart: PermissionGroup.microphone
                # 'PERMISSION_MICROPHONE=1',
        
                ## dart: PermissionGroup.speech
                # 'PERMISSION_SPEECH_RECOGNIZER=1',
        
                ## dart: PermissionGroup.photos
                # 'PERMISSION_PHOTOS=1',
        
                ## dart: [PermissionGroup.location, PermissionGroup.locationAlways, PermissionGroup.locationWhenInUse]
                # 'PERMISSION_LOCATION=1',
        
                ## dart: PermissionGroup.notification
                # 'PERMISSION_NOTIFICATIONS=1',
        
                ## dart: PermissionGroup.mediaLibrary
                # 'PERMISSION_MEDIA_LIBRARY=1',
        
                ## dart: PermissionGroup.sensors
                # 'PERMISSION_SENSORS=1',   
        
                ## dart: PermissionGroup.bluetooth
                # 'PERMISSION_BLUETOOTH=1',
        
                ## dart: PermissionGroup.appTrackingTransparency
                # 'PERMISSION_APP_TRACKING_TRANSPARENCY=1',
        
                ## dart: PermissionGroup.criticalAlerts
                # 'PERMISSION_CRITICAL_ALERTS=1'
              ]
        
            end
          end
        end
        
    
    copied to clipboard
    
2.  Remove the `#` character in front of the permission you do want to use. For example if you need access to the calendar make sure the code looks like this:
    
                ## dart: PermissionGroup.calendar
                'PERMISSION_EVENTS=1',
        
    
    copied to clipboard
    
3.  Delete the corresponding permission description in `Info.plist` e.g. when you don't need camera permission, just delete 'NSCameraUsageDescription' The following lists the relationship between `Permission` and `The key of Info.plist`:
    
    Permission
    
    Info.plist
    
    Macro
    
    PermissionGroup.calendar
    
    NSCalendarsUsageDescription
    
    PERMISSION\_EVENTS
    
    PermissionGroup.reminders
    
    NSRemindersUsageDescription
    
    PERMISSION\_REMINDERS
    
    PermissionGroup.contacts
    
    NSContactsUsageDescription
    
    PERMISSION\_CONTACTS
    
    PermissionGroup.camera
    
    NSCameraUsageDescription
    
    PERMISSION\_CAMERA
    
    PermissionGroup.microphone
    
    NSMicrophoneUsageDescription
    
    PERMISSION\_MICROPHONE
    
    PermissionGroup.speech
    
    NSSpeechRecognitionUsageDescription
    
    PERMISSION\_SPEECH\_RECOGNIZER
    
    PermissionGroup.photos
    
    NSPhotoLibraryUsageDescription
    
    PERMISSION\_PHOTOS
    
    PermissionGroup.location, PermissionGroup.locationAlways, PermissionGroup.locationWhenInUse
    
    NSLocationUsageDescription, NSLocationAlwaysAndWhenInUseUsageDescription, NSLocationWhenInUseUsageDescription
    
    PERMISSION\_LOCATION
    
    PermissionGroup.notification
    
    PermissionGroupNotification
    
    PERMISSION\_NOTIFICATIONS
    
    PermissionGroup.mediaLibrary
    
    NSAppleMusicUsageDescription, kTCCServiceMediaLibrary
    
    PERMISSION\_MEDIA\_LIBRARY
    
    PermissionGroup.sensors
    
    NSMotionUsageDescription
    
    PERMISSION\_SENSORS
    
    PermissionGroup.bluetooth
    
    NSBluetoothAlwaysUsageDescription, NSBluetoothPeripheralUsageDescription
    
    PERMISSION\_BLUETOOTH
    
    PermissionGroup.appTrackingTransparency
    
    NSUserTrackingUsageDescription
    
    PERMISSION\_APP\_TRACKING\_TRANSPARENCY
    
    PermissionGroup.criticalAlerts
    
    PermissionGroupCriticalAlerts
    
    PERMISSION\_CRITICAL\_ALERTS
    
4.  Clean & Rebuild
</details>

***


This plugin is suitable for iOS and Android. This plugin allows Flutter apps to detect network changes. You can know the detailed mobile network types, such as 2G, 3G, 4G, 5G.
This plugin also allows you to see the status of internet connection

## Use

#### Example

![alt text](https://media.giphy.com/media/Q38gPMCyFWWJZVj8g0/giphy-downsized-large.gif "Network_type_reachability")


1. get NETWORK_TYPE
**
Note: Android must dynamically obtain the READ_PHONE_STATE permission to judge 2G/3G/4G/5G**
```dart


  String _networkTypeStatic = 'Unknown';

  //widget
  Column(
    mainAxisAlignment: MainAxisAlignment.spaceAround,
    children: [
      const Text('NETWORK_TYPE : '),
      Text(
        _networkTypeStatic,
        style: const TextStyle(fontWeight: FontWeight.bold),
      ),
      TextButton(
        onPressed: () {
          _getCurrentNetworkStatus();
        },
        child: const Text('Get-Data'),
        style: ButtonStyle(
          backgroundColor: MaterialStateProperty.all(
              const Color.fromARGB(255, 162, 229, 188)),
        ),
      )
    ],
  ),

  //function

  _getCurrentNetworkStatus() async {
    if (Platform.isAndroid) {
      await NetworkTypeReachability().getPermisionsAndroid;
    }
    NetworkStatus status =
        await NetworkTypeReachability().currentNetworkStatus();
    switch (status) {
      case NetworkStatus.unreachable:
      //unreachable
      case NetworkStatus.wifi:
      //wifi
      case NetworkStatus.mobile2G:
      //2g
      case NetworkStatus.mobile3G:
      //3g
      case NetworkStatus.mobile4G:
      //4g
      case NetworkStatus.mobile5G:
      //5h
      case NetworkStatus.othermobile:
      //other
    }
    setState(() {
      _networkTypeStatic = status.toString();
    });
  }
    
```
2. You can also listen for NETWORK_TYPE changes by subscribing to the stream exposed by Network_type_reachability plugin
```dart

  String _networkTypeSuscription = 'Unknown';
  StreamSubscription<NetworkStatus>? subscriptionNetworkType;

  @override
  void initState() {
    super.initState();
    _listenNetworkStatus();
  }

  @override
  void dispose() {
    super.dispose();
    subscriptionNetworkType?.cancel();
  }

  _listenNetworkStatus() async {
    if (Platform.isAndroid) {
      await NetworkTypeReachability().getPermisionsAndroid;
    }
    subscriptionNetworkType =
        NetworkTypeReachability().onNetworkStateChanged.listen((event) {
      setState(() {
        _networkTypeSuscription = "$event";
      });
    });
  }


  // Widget
  Column(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
      const Text('NETWORK_TYPE Suscription: '),
      Text(
        _networkTypeSuscription,
        style: const TextStyle(fontWeight: FontWeight.bold),
      ),
    ],
  ),
```
3. get Status Internet Conection 

```dart

  String connectivityInternetStatic = 'Unknown';

  // Widget
  Column(
    mainAxisAlignment: MainAxisAlignment.spaceAround,
    children: [
      const Text('Status Internet Conection : '),
      Text(
        connectivityInternetStatic,
        style: const TextStyle(fontWeight: FontWeight.bold),
      ),
      TextButton(
        onPressed: () async {
          print('#=======> cargando');
          connectivityInternetStatic = 'loading...';
          setState(() {});
          InternetStatusConnection data =
              await NetworkTypeReachability()
                  .getInternetStatusConnection();
          print(data);
          print('#=======> finalizando');

          connectivityInternetStatic = data.toString();
          setState(() {});
        },
        child: const Text('Get-Data'),
        style: ButtonStyle(
          backgroundColor: MaterialStateProperty.all(
              const Color.fromARGB(255, 162, 229, 188)),
        ),
      )
    ],
  ),
    
```
4. get Status Internet Conection Suscription
NOTE: It is recommended not to keep the listening state constantly since behind what it does is load the google.com page and this would constantly consume many megabytes of internet over time
```dart

  String connectivityInternetSuscription = 'Unknown';
  StreamSubscription<InternetStatusConnection>? subscriptionInternetConnection;

  @override
  void initState() {
    super.initState();
    _listenInternetConnection();
  }

  @override
  void dispose() {
    super.dispose();
    subscriptionInternetConnection?.cancel();
    NetworkTypeReachability().listenInternetConnection = false;
  }


  // Widget
  Column(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
      const Text('Status Internet Conection : '),
      Text(
        connectivityInternetSuscription,
        style: const TextStyle(fontWeight: FontWeight.bold),
      ),
    ],
  )


  _listenInternetConnection() async {
    subscriptionInternetConnection = NetworkTypeReachability()
        .getStreamInternetConnection(showLogs: false)
        .listen((event) {
      setState(() {
        connectivityInternetSuscription = event.toString();
      });
    });
  }

```

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 1.3.3 - getInternetStatusConnection
* Add description
## 1.3.2 - Add description
* Add description
## 1.3.1 - NullSafety code
* Support NullSafety code example
## 1.3.0 - NullSafety
* Support NullSafety
## 1.2.1 - recommendation
* to not consume a lot of mobile data
## 1.2.0 - Add Internet Connection
* Add Internet Connection
## 1.0.1 - changes
* ios change and suport from 14.0 to 14.1
## 1.0.0

* TODO: Describe initial release.


lib/network_type_reachability.dart
import 'dart:async';
import 'dart:developer';

import 'package:flutter/services.dart';

import 'package:dart_ping/dart_ping.dart';
import 'package:dart_ping_ios/dart_ping_ios.dart';

enum NetworkStatus {
  unreachable,
  wifi,
  mobile2G,
  mobile3G,
  mobile4G,
  mobile5G,
  otherMobile
}

enum InternetStatusConnection {
  withoutInternet,
  withInternet,
  unstableInternet,
}

class NetworkTypeReachability {
  // make this nullable by adding '?'
  static NetworkTypeReachability? _instance;

  factory NetworkTypeReachability() {
    DartPingIOS.register();
    _instance ??= NetworkTypeReachability._();
    return _instance!;
  }
  NetworkTypeReachability._();

  static const MethodChannel _channel =
      MethodChannel('flutter_plugin_reachability');

  static const EventChannel _eventChannel =
      EventChannel("flutter_plugin_reachability_status");

  Stream<NetworkStatus>? _onNetworkStateChanged;

  /// currentNetworkStatus obtain the status network in live
  Stream<NetworkStatus> get onNetworkStateChanged {
    _onNetworkStateChanged ??= _eventChannel
        .receiveBroadcastStream()
        .map((event) => event.toString())
        .map(_convertFromState);
    return _onNetworkStateChanged!;
  }

  /// currentNetworkStatus obtain the status network
  Future<NetworkStatus> currentNetworkStatus() async {
    final String state = await _channel.invokeMethod("networkStatus");
    return _convertFromState(state);
  }

  /// NetworkStatus identify tipe of connection
  NetworkStatus _convertFromState(String state) {
    switch (state) {
      case "unreach":
        return NetworkStatus.unreachable;
      case "mobile2G":
        return NetworkStatus.mobile2G;
      case "mobile3G":
        return NetworkStatus.mobile3G;
      case "wifi":
        return NetworkStatus.wifi;
      case "mobile4G":
        return NetworkStatus.mobile4G;
      case "mobile5G":
        return NetworkStatus.mobile5G;
      case "mobileOther":
        return NetworkStatus.otherMobile;
      default:
        return NetworkStatus.unreachable;
    }
  }

  /// performs a sending and receiving of packets to an internet page,
  /// if the number of packets sent is equal to the number of packets received
  /// then if there is a good internet connection
  Future<InternetStatusConnection> getInternetStatusConnection({
    urlTest = 'google.com',
    countPing = 3,
    timeOutIntents = 5,
    showLogs = false,
  }) async {
    NetworkStatus networkStatus = await currentNetworkStatus();
    if (networkStatus == NetworkStatus.unreachable) {
      return InternetStatusConnection.withoutInternet;
    }
    Ping ping = Ping(urlTest, count: countPing);
    PingData pingData = await ping.stream.last
        .timeout(
      Duration(seconds: timeOutIntents),
    )
        .catchError((e) {
      return throw e!;
    }).onError((error, stackTrace) {
      return throw error!;
    });
    if (showLogs) {
      log('Running PING ===== > $pingData');
    }
    try {
      if (pingData.summary?.transmitted == pingData.summary?.received) {
        return InternetStatusConnection.withInternet;
      } else if (pingData.summary!.transmitted > 0 &&
          pingData.summary!.received > 0) {
        return InternetStatusConnection.unstableInternet;
      } else {
        return InternetStatusConnection.withoutInternet;
      }
    } catch (e) {
      return InternetStatusConnection.withoutInternet;
    }
  }

  /// from here on the code focuses on maintaining a listening state wondering if there is an internet connection or not
  bool listenInternet = true;
  bool get listenInternetConnection => listenInternet;
  set listenInternetConnection(data) {
    listenInternet = data;
  }

  Stream<InternetStatusConnection> getStreamInternetConnection(
      {showLogs = false}) async* {
    InternetStatusConnection? globalStatusConnection;
    while (listenInternetConnection) {
      try {
        InternetStatusConnection statusConnection =
            await getInternetStatusConnection(showLogs: showLogs);
        if (globalStatusConnection != statusConnection) {
          globalStatusConnection = statusConnection;
          yield globalStatusConnection;
        }
      } catch (error) {
        log(error);
      }
      await Future.delayed(const Duration(seconds: 5));
    }
  }
}


ios/Classes/SwiftNetworkTypeReachabilityPlugin.swift
import Flutter
import UIKit
import Reachability
import CoreTelephony

public class SwiftNetworkTypeReachabilityPlugin: NSObject, FlutterPlugin {

    private var reachability: Reachability?
    private var eventSink: FlutterEventSink?
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_plugin_reachability", binaryMessenger: registrar.messenger())
        let instance = SwiftNetworkTypeReachabilityPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        let streamChannel = FlutterEventChannel(name: "flutter_plugin_reachability_status", binaryMessenger: registrar.messenger())
        streamChannel.setStreamHandler(instance)

    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "networkStatus" {
            result(statusForNetWork())
        }else{
            result(FlutterMethodNotImplemented)
        }
    }
}


extension SwiftNetworkTypeReachabilityPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        //1添加通知
        NotificationCenter.default.addObserver(self, selector:#selector(networkStatusChange(_:)) , name: .reachabilityChanged, object: nil)
        do {
            reachability = try Reachability()
            try reachability?.startNotifier()
        }catch(let erro as NSError){
            return FlutterError(code: "\(erro.code)", message: erro.domain, details: "初始化网络监测失败");
        }
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        if reachability != nil {
            do {
                try reachability?.startNotifier()
                reachability = nil
            }catch(let erro as NSError){
                return FlutterError(code: "\(erro.code)", message: erro.domain, details: "停止网络监测失败");
            }
        }
        NotificationCenter.default.removeObserver(self)
        return nil
    }


    @objc func networkStatusChange(_ notice: Notification) {
        eventSink?(statusForNetWork())
    }

    private func statusForNetWork() -> String {
        //0:unreachable 1:2G 2:3G 3:Wi-Fi 4:4G 5:5G 6:othermobie
        if reachability == nil {
             reachability = try! Reachability()
        }
        guard let netReach = reachability else { return "0" }
        switch netReach.connection {
        case .wifi:
            return NetworkStatus.wifi.value
        case .unavailable:
            return NetworkStatus.unreach.value
        case .cellular:

            if Float(String(UIDevice.current.systemVersion.split(separator: ".").first!)) ?? 0 >= 7.0 {
                let teleInfo = CTTelephonyNetworkInfo()
                guard let access = teleInfo.currentRadioAccessTechnology else {
                    return  NetworkStatus.other.value
                }
                if [CTRadioAccessTechnologyEdge,
                    CTRadioAccessTechnologyGPRS,
                    CTRadioAccessTechnologyCDMA1x].contains(access) {
                    return NetworkStatus.mobile2G.value
                }

                if [CTRadioAccessTechnologyHSDPA,
                    CTRadioAccessTechnologyWCDMA,
                    CTRadioAccessTechnologyHSUPA,
                    CTRadioAccessTechnologyCDMAEVDORev0,
                    CTRadioAccessTechnologyCDMAEVDORevA,
                    CTRadioAccessTechnologyCDMAEVDORevB,
                    CTRadioAccessTechnologyeHRPD].contains(access) {
                    return NetworkStatus.mobile3G.value
                }

                if [CTRadioAccessTechnologyLTE].contains(access) {
                    return NetworkStatus.mobile4G.value
                }
                if #available(iOS 14.1, *), [CTRadioAccessTechnologyNRNSA,
                                             CTRadioAccessTechnologyNR].contains(access) {
                    return NetworkStatus.mobile5G.value
                }
                return NetworkStatus.other.value

            }else{
                return NetworkStatus.other.value
            }
        default:
            return NetworkStatus.unreach.value
        }
    }

    enum NetworkStatus: String {
        case unreach = "unreach"
        case mobile2G = "mobile2G"
        case mobile3G = "mobile3G"
        case wifi = "wifi"
        case mobile4G = "mobile4G"
        case mobile5G = "mobile5G"
        case other = "mobileOther"
        var value: String{
            return "\(self.rawValue)"
        }
    }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:network_type_reachability/network_type_reachability.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _networkTypeStatic = 'Unknown';

  String _networkTypeSuscription = 'Unknown';
  StreamSubscription<NetworkStatus>? subscriptionNetworkType;

  String connectivityInternetStatic = 'Unknown';

  String connectivityInternetSuscription = 'Unknown';
  StreamSubscription<InternetStatusConnection>? subscriptionInternetConnection;

  @override
  void initState() {
    super.initState();
    _listenNetworkStatus();
    _listenInternetConnection();
  }

  _listenNetworkStatus() async {
    subscriptionNetworkType =
        NetworkTypeReachability().onNetworkStateChanged.listen((event) {
      setState(() {
        _networkTypeSuscription = "$event";
      });
    });
  }

  _listenInternetConnection() async {
    subscriptionInternetConnection = NetworkTypeReachability()
        .getStreamInternetConnection(showLogs: false)
        .listen((event) {
      setState(() {
        connectivityInternetSuscription = event.toString();
      });
    });
  }

  _getCurrentNetworkStatus() async {
    NetworkStatus status =
        await NetworkTypeReachability().currentNetworkStatus();
    switch (status) {
      case NetworkStatus.unreachable:
      //unreachable
      case NetworkStatus.wifi:
      //wifi
      case NetworkStatus.mobile2G:
      //2g
      case NetworkStatus.mobile3G:
      //3g
      case NetworkStatus.mobile4G:
      //4g
      case NetworkStatus.mobile5G:
      //5h
      case NetworkStatus.otherMobile:
      //other
    }
    setState(() {
      _networkTypeStatic = status.toString();
    });
  }

  @override
  void dispose() {
    super.dispose();
    subscriptionNetworkType?.cancel();
    subscriptionInternetConnection?.cancel();
    NetworkTypeReachability().listenInternetConnection = false;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Network Type Reachability'),
          backgroundColor: Colors.blueAccent[900],
        ),
        body: Column(
          children: [
            _expandedContainerRow(
              flex: 1,
              color: Colors.blueGrey[900],
              children: [
                Expanded(
                  child: _box(
                    child: const Center(
                      child: Text(
                        'Static Data',
                        style: TextStyle(
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                          fontSize: 20,
                        ),
                      ),
                    ),
                  ),
                ),
                Expanded(
                  child: _box(
                    child: const Center(
                      child: Text(
                        'Listen to changes',
                        style: TextStyle(
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                          fontSize: 20,
                        ),
                      ),
                    ),
                  ),
                ),
              ],
            ),
            _expandedContainerRow(
              children: [
                Expanded(
                  child: _box(
                    color: _colorStatusNetworkType(_networkTypeStatic),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.spaceAround,
                      children: [
                        const Text('NETWORK_TYPE : '),
                        Text(
                          _networkTypeStatic,
                          style: const TextStyle(fontWeight: FontWeight.bold),
                        ),
                        TextButton(
                          onPressed: () {
                            _getCurrentNetworkStatus();
                          },
                          child: const Text('Get-Data'),
                          style: ButtonStyle(
                            backgroundColor: MaterialStateProperty.all(
                                const Color.fromARGB(255, 162, 229, 188)),
                          ),
                        )
                      ],
                    ),
                  ),
                ),
                Expanded(
                  child: _box(
                    color: _colorStatusNetworkType(_networkTypeSuscription),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        const Text('NETWORK_TYPE Suscription: '),
                        Text(
                          _networkTypeSuscription,
                          style: const TextStyle(fontWeight: FontWeight.bold),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
            _expandedContainerRow(
              children: [
                Expanded(
                  child: _box(
                    color: _colorStatusInternetType(connectivityInternetStatic),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.spaceAround,
                      children: [
                        const Text('Status Internet Conection : '),
                        Text(
                          connectivityInternetStatic,
                          style: const TextStyle(fontWeight: FontWeight.bold),
                        ),
                        TextButton(
                          onPressed: () async {
                            print('#=======> cargando');
                            connectivityInternetStatic = 'loading...';
                            setState(() {});
                            InternetStatusConnection data =
                                await NetworkTypeReachability()
                                    .getInternetStatusConnection();
                            print(data);
                            print('#=======> finalizando');

                            connectivityInternetStatic = data.toString();
                            setState(() {});
                          },
                          child: const Text('Get-Data'),
                          style: ButtonStyle(
                            backgroundColor: MaterialStateProperty.all(
                                const Color.fromARGB(255, 162, 229, 188)),
                          ),
                        )
                      ],
                    ),
                  ),
                ),
                Expanded(
                  child: _box(
                    color: _colorStatusInternetType(
                        connectivityInternetSuscription),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        const Text('Status Internet Conection : '),
                        Text(
                          connectivityInternetSuscription,
                          style: const TextStyle(fontWeight: FontWeight.bold),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }

  Expanded _expandedContainerRow({
    required List<Widget> children,
    Color? color = Colors.transparent,
    int flex = 4,
  }) {
    return Expanded(
      flex: flex,
      child: Container(
        color: color,
        child: Row(
          children: children,
        ),
      ),
    );
  }

  Container _box({
    required Widget child,
    Color color = Colors.transparent,
  }) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
      decoration: BoxDecoration(
        border: Border.all(color: Colors.black26),
        color: color,
      ),
      height: double.infinity,
      child: child,
    );
  }

  Color _colorStatusNetworkType(String data) {
    // case NetworkStatus.unreachable:
    // case NetworkStatus.wifi:
    // case NetworkStatus.mobile2G:
    // case NetworkStatus.mobile3G:
    // case NetworkStatus.mobile4G:
    // case NetworkStatus.mobile5G:
    // case NetworkStatus.othermobile:
    switch (data) {
      case 'Unknown':
        return Colors.yellow.shade100;
      case 'NetworkStatus.unreachable':
        return Colors.red.shade100;
      default:
        return Colors.green.shade100;
    }
  }

  Color _colorStatusInternetType(String data) {
    // withoutInternet,
    // withInternet,
    // unstableInternet,
    switch (data) {
      case 'Unknown':
        return Colors.yellow.shade100;
      case 'InternetStatusConnection.withoutInternet':
        return Colors.red.shade100;
      case 'InternetStatusConnection.withInternet':
        return Colors.green.shade100;
      case 'InternetStatusConnection.unstableInternet':
        return Colors.brown.shade100;
      default:
        return Colors.yellow.shade100;
    }
  }
}
